nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -1266,6 +1279,46 @@ public static boolean isBooleanTypeCompatible(final Object value) {
         return false;
     }
 
+    public static BigDecimal toBigDecimal(final Object value, final String fieldName) {
+        if (value == null) {
+            return null;
+        }
+
+        if (value instanceof BigDecimal) {
+            return (BigDecimal) value;
+        }
+
+        if (value instanceof Number) {
+            final Number number = (Number) value;
+
+            if (number instanceof Byte
+                    || number instanceof Short
+                    || number instanceof Integer
+                    || number instanceof Long) {
+                return BigDecimal.valueOf(number.longValue());
+            }
+
+            if (number instanceof BigInteger) {
+                return new BigDecimal((BigInteger) number);
+            }
+
+            if (number instanceof Float || number instanceof Double) {
+                return BigDecimal.valueOf(number.doubleValue());
```

Comment:
`BigDecimal.valueOf(double)` seems to me fine.
It calls `new BigDecimal(Double.toString(val))` in the background, which is the proposed solution.
I think the problematic way would be the `new BigDecimal(double)` constructor, not `BigDecimal.valueOf(double)`.

`new BigDecimal(number.toString())` could also be used.

Reply:
I did update as it is not worse than the original one, but honestly I could not reproduce any issue with that (behaves the same way for NaN and Infinity). However, I cannot close out a situation where it might happen so it is better be on the safe side.